### PR TITLE
fuzz: link libfuzzer engine to fuzz with bazel!

### PR DESCRIPTION
**Pull request by asraa**
This makes a new config "asan-fuzzer" that links to the libfuzzer engine.

To run a fuzz target with asan and libfuzzer, do
bazel build --config=asan-fuzzer test/common/router:route_fuzz_test_driver --config=asan-fuzzer
bazel-bin/test/common/router/route_fuzz_test_driver test/common/router/route_corpus -runs=-1

(-runs=-1 for indefinite runs, -runs=100 for bounded).

Testing: works locally
Risk: Low

### CODE REVIEW

File: bazel/envoy_test.bzl [hunk 1]
```diff
@@ -115,6 +115,16 @@ def envoy_cc_fuzz_test(name, corpus, deps = [], tags = [], **kwargs):
         tags = ["manual"] + tags,
     )
 
+    native.cc_binary(
+        name = name + "_driver",
```
Comment (htuch) [inline bazel/envoy_test.bzl]: Nit: I would call this `_with_libfuzzer` rather than `_driver`.

**Comment by yevgenypats** [comment]
This is awesome, I'll rebase my PR on top of this one, once it is merged!.

**Comment by yevgenypats** [comment]
looks like one of tests here fails. maybe it's out of memory?

**Comment by lizan** [comment]
Can you try applying this? I think it will fix clang_tidy:

```
diff --git a/tools/gen_compilation_database.py b/tools/gen_compilation_database.py
index 94869312b..4090b0505 100755
--- a/tools/gen_compilation_database.py
+++ b/tools/gen_compilation_database.py
@@ -8,7 +8,7 @@ import subprocess

 def generateCompilationDatabase(args):
   if args.run_bazel_build:
-    subprocess.check_call(["bazel", "build"] + args.bazel_targets)
+    subprocess.check_call(["bazel", "build", "--build_tag_filters=-manual"] + args.bazel_targets)

   gen_compilation_database_sh = os.path.join(
       os.path.realpath(os.path.dirname(__file__)), "../bazel/gen_compilation_database.sh")
```

Also merge master to pick up CI changes.

**Comment by yevgenypats** [comment]
This is still gets killed due to oom. I suspect that you need to add `bazel build --jobs 8` or otherwise it will use 200 workers and will overuse the memory and will get killed by Circle.

**Comment by asraa** [comment]
Thank you! Yes, let's see. I'm not sure what the implications are later, but I did add the jobs flag to see what happens.

**Comment by yevgenypats** [comment]
horray! it worked:) #why_bazel_spins_up_200_workers_by_default:)

**Comment by asraa** [comment]
Yes! 

@htuch, @lizan -- what do you think about this? (if this merges, probably everyone will OOM CI without the --jobs flag. how many jobs are appropriate, or is there a better way to mitigate this?)

**Comment by lizan** [comment]
Can you try to fix the jobs number to env ${NUM_CPUS} (which is set to 8 in CI)?

@yevgenypats bazel automatically spins up job to the number of CPU cores, which ignores cgroup, so in a docker environment in large host, it spins job more than cgroup restriction.

**Comment by asraa** [comment]
Thanks everyone. clang-tidy runs with NUM_CPUS jobs now.

**Comment by yevgenypats** [comment]
horray! lets merge!

**Pull request merged**